src/server/database/database_provider.py

```diff
@@ -0,0 +1,113 @@
+                name: list[str] = response.data
+            else:
+                # handle query failure
+    """
+
+    connection: sqlite3.Connection = sqlite3.Connection(":memory:")
+    database_source: str = ""
+
+    @classmethod
+    def initialize(cls) -> None:
+        """Initialize connection to the database."""
+        cls.database_source = PATHS.DATABASE
+
+        if not os.path.exists(cls.database_source):
+            result = cls._create_database()
+            if result is Message.NO_CONNECTION:
+                print("INFO: connecting to in-memory database as fallback")
+                cls.database_source = ":memory:"
+                cls._create_database()
+        cls._fill_database()
+
+    @classmethod
+    @contextmanager
+    def query(
+        cls, query: str, params: tuple | list = ()
+    ) -> Generator[DatabaseResponse, None, None]:
+        """Interface that allows safe query execution in database.
+
+        Args:
+            query (str): Query to be executed in the database.
+            params (tuple, optional): Parameters passed to sqlite3 query string. Defaults to ().
+
+        Yields:
+            DatabaseResponse: Batch containing query execution status and received data (if any).
+
+        """
+        with cls._try_get_cursor() as cursor:
+            if cursor is None:
+                yield DatabaseResponse(Message.NO_CONNECTION, [])
+            else:
+                try:
+                    print(f"INFO: {query=}, {params=}")
+                    cursor.execute(query, params)
+                    data: list = cursor.fetchall()
+                    yield DatabaseResponse(Message.OK, data)
+                except sqlite3.IntegrityError as err:
+                    print(f"ERROR: {err}")
+                    yield DatabaseResponse(Message.INVALID_SQL, [])
+                except sqlite3.ProgrammingError as err:
+                    print(f"ERROR: {err}")
+                    yield DatabaseResponse(Message.INVALID_BINDINGS, [])
+                except sqlite3.Error as err:
+                    print(f"ERROR: {err}")
+                    yield DatabaseResponse(Message.UNKNOWN_ERROR, [])
+                finally:
+                    print(f"INFO: successfully executed a query: '{query}'")
+
+    @classmethod
+    @contextmanager
+    def _try_get_cursor(cls) -> Generator[Optional[sqlite3.Cursor], None, None]:
+        cursor: Optional[sqlite3.Cursor] = None
+        try:
+            cursor = cls.connection.cursor()
+            yield cursor
+        except sqlite3.Error as err:
+            print(f"ERROR: {err}")
+            yield None
+        finally:
+            if cursor is not None:
+                cursor.close()
+                cls.connection.commit()
+
+    @classmethod
+    def _create_database(cls) -> Message:
+        try:
+            print(f"INFO: {cls.database_source=}")
+            cls.connection = sqlite3.connect(cls.database_source, check_same_thread=False)
```

Comment:
Z tego co widzę, to jedynie miejsce, gdzie używa się cls.database_source do utworzenia połączenia.
Czyli poza pierwszym uruchomieniem, które tworzy bazę danych, pozostałe połączenia są wyłącznie w pamięci. 
Nie wywala podczas testów, bo _fill_database jest wywoływane za każdym razem

Reply:
I don't understand what is the issue here, can you elaborate?